langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java

```diff
@@ -0,0 +1,244 @@
+    }
+
+    @Override
+    public String add(Embedding embedding, TextSegment textSegment) {
+        String id = randomUUID();
+        addInternal(id, embedding, textSegment);
+        return id;
+    }
+
+    @Override
+    public List<String> addAll(List<Embedding> embeddings) {
+        return addAll(embeddings, null);
+    }
+
+    @Override
+    public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedded) {
+        List<String> ids = embeddings.stream()
+                .map(ignored -> randomUUID())
+                .toList();
+        addAllInternal(ids, embeddings, embedded);
+        return ids;
+    }
+
+    @Override
+    public List<EmbeddingMatch<TextSegment>> findRelevant(Embedding referenceEmbedding, int maxResults, double minScore) {
+        var embeddingValue = Values.value(referenceEmbedding.vector());
+        
+        try (var session = session()) {
+            return session
+                    .run("""
+						CALL db.index.vector.queryNodes($indexName, $maxResults, $embeddingValue)
+						YIELD node, score
+						WHERE score >= $minScore
+						RETURN node, score
+						""", Map.of("indexName", indexName,
+                            "embeddingValue", embeddingValue,
+                            "minScore", minScore,
+                            "maxResults", maxResults))
+                    .list(item -> Neo4jEmbeddingUtils.toEmbeddingMatch(this, item));
+        }
+    }
+
+
+    /*
+    Private methods
+    */
+    
+    private void addInternal(String id, Embedding embedding, TextSegment embedded) {
+        addAllInternal(singletonList(id), singletonList(embedding), embedded == null ? null : singletonList(embedded));
+    }
+
+    private void addAllInternal(List<String> ids, List<Embedding> embeddings, List<TextSegment> embedded) {
+        if (isCollectionEmpty(ids) || isCollectionEmpty(embeddings)) {
+            log.info("[do not add empty embeddings to neo4j]");
+            return;
+        }
+        ensureTrue(ids.size() == embeddings.size(), "ids size is not equal to embeddings size");
+        ensureTrue(embedded == null || embeddings.size() == embedded.size(), "embeddings size is not equal to embedded size");
+
+        bulk(ids, embeddings, embedded);
+    }
+
+    private void bulk(List<String> ids, List<Embedding> embeddings, List<TextSegment> embedded) {
+
+        List<Map<String, Object>> rows = IntStream.range(0, ids.size())
+                .mapToObj(idx -> toRecord(this, idx, ids, embeddings, embedded))
+                .toList();
+
+        try (var session = session()) {
+            String statement = """
+                            UNWIND $rows AS row
+                            MERGE (u:%1$s {%2$s: row.%2$s})
+                            SET u += row.%3$s
+                            WITH row, u
+                            CALL db.create.setNodeVectorProperty(u, $embeddingProperty, row.%4$s)
+                            RETURN *""".formatted(
```

Comment:
we should return count(*) to avoid huge results coming back (think of all the input data and all the nodes with their properties coming back).